server_udp.go

```diff
@@ -42,83 +47,152 @@ func (w *UDPResponseWriter) WriteMsg(m *dns.Msg) error {
 		log.Tracef("Failed to encrypt the DNS query: %v", err)
 		return err
 	}
-
-	_, _ = udpWrite(res, w.udpConn, w.remoteAddr, w.localIP)
-	return nil
+	_, err = dns.WriteToSessionUDP(w.udpConn, res, w.sess)
+	return err
 }
 
 // ServeUDP - listens to UDP connections, queries are then processed by Server.Handler.
-// It blocks the calling goroutine and to stop it you need to close the listener.
+// It blocks the calling goroutine and to stop it you need to close the listener
+// or call Server.Shutdown.
 func (s *Server) ServeUDP(l *net.UDPConn) error {
+	var once sync.Once
+	unlock := func() { once.Do(s.lock.Unlock) }
```

Comment:
Such constructs usually mean that you need to factor out the functionality into a separate function and just use `lock(); defer unlock()` there.  The methods, this one and the TCP one, are way too complex, imo.

Reply:
They definitely are too complex